{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Laboratorio 9: GEO – Gene Expression Omnibus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pregunta A:\n",
    "\n",
    "La finalidad de *GEO2R* es poder identificar genes que sean expresados de manera diferenciada comparando dos o más muestras de una *GEO Series* a lo largo de condiciones experimentales. El resultado es brindado como una tabla de genes en orden de importancia."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preguntas B-K:\n",
    "\n",
    "Al momento de insertar el número de acceso, obtenemos los resultados dados por la base de datos\n",
    "\n",
    "![](GSE18388.png)\n",
    "\n",
    "Luego podemos definir grupos para comparar las muestras, deben ser al menos 2 grupos cada uno con al menos 1 muestra. En nuestro caso definimos los grupos `prueba` (celeste) y `control` (rosa).\n",
    "\n",
    "![](SeteandoGrupos.png)\n",
    "\n",
    "Al presionar el botón **Top 250** aparecen los primeros 250 genes relacionados.\n",
    "\n",
    "![](ResultadosTabla.png)\n",
    "\n",
    "Cuando se abre la pestaña de alguno de los genes aparece un gráfico de comparación:\n",
    "\n",
    "![](ResultadosIndividuales.png)\n",
    "\n",
    "De la tabla, notamos que la mayoría de genes tienen un valor-P cercano entre sí, el rango general del valor-P sin procesar es $[0.0000804,\t0.01468]$.\n",
    "\n",
    "Los siguientes son los resultados de los primeros 4 genes:\n",
    "\n",
    "![Gen1](ResultadosGen1.png)\n",
    "\n",
    "![Gen2](ResultadosGen2.png)\n",
    "\n",
    "![Gen3](ResultadosGen3.png)\n",
    "\n",
    "![Gen4](ResultadosGen4.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "El R-Script dado es el siguiente:\n",
    "\n",
    "```R\n",
    "# Version info: R 3.2.3, Biobase 2.30.0, GEOquery 2.40.0, limma 3.26.8\n",
    "# R scripts generated  Fri Nov 10 10:44:26 EST 2017\n",
    "\n",
    "################################################################\n",
    "#   Differential expression analysis with limma\n",
    "library(Biobase)\n",
    "library(GEOquery)\n",
    "library(limma)\n",
    "\n",
    "# load series and platform data from GEO\n",
    "\n",
    "gset <- getGEO(\"GSE18388\", GSEMatrix =TRUE, AnnotGPL=TRUE)\n",
    "if (length(gset) > 1) idx <- grep(\"GPL6246\", attr(gset, \"names\")) else idx <- 1\n",
    "gset <- gset[[idx]]\n",
    "\n",
    "# make proper column names to match toptable \n",
    "fvarLabels(gset) <- make.names(fvarLabels(gset))\n",
    "\n",
    "# group names for all samples\n",
    "gsms <- \"10101010\"\n",
    "sml <- c()\n",
    "for (i in 1:nchar(gsms)) { sml[i] <- substr(gsms,i,i) }\n",
    "\n",
    "# log2 transform\n",
    "ex <- exprs(gset)\n",
    "qx <- as.numeric(quantile(ex, c(0., 0.25, 0.5, 0.75, 0.99, 1.0), na.rm=T))\n",
    "LogC <- (qx[5] > 100) ||\n",
    "          (qx[6]-qx[1] > 50 && qx[2] > 0) ||\n",
    "\t            (qx[2] > 0 && qx[2] < 1 && qx[4] > 1 && qx[4] < 2)\n",
    "if (LogC) { ex[which(ex <= 0)] <- NaN\n",
    "\t\t      exprs(gset) <- log2(ex) }\n",
    "\n",
    "# set up the data and proceed with analysis\n",
    "sml <- paste(\"G\", sml, sep=\"\")    # set group names\n",
    "fl <- as.factor(sml)\n",
    "gset$description <- fl\n",
    "design <- model.matrix(~ description + 0, gset)\n",
    "colnames(design) <- levels(fl)\n",
    "fit <- lmFit(gset, design)\n",
    "cont.matrix <- makeContrasts(G1-G0, levels=design)\n",
    "fit2 <- contrasts.fit(fit, cont.matrix)\n",
    "fit2 <- eBayes(fit2, 0.01)\n",
    "tT <- topTable(fit2, adjust=\"fdr\", sort.by=\"B\", number=250)\n",
    "tT <- subset(tT, select=c(\"ID\",\"adj.P.Val\",\"P.Value\",\"t\",\"B\",\"logFC\",\"Gene.symbol\",\"Gene.title\"))\n",
    "write.table(tT, file=stdout(), row.names=F, sep=\"\\t\")\n",
    "\n",
    "################################################################\n",
    "#   Boxplot for selected GEO samples\n",
    "library(Biobase)\n",
    "library(GEOquery)\n",
    "\n",
    "# load series and platform data from GEO\n",
    "\n",
    "gset <- getGEO(\"GSE18388\", GSEMatrix =TRUE, getGPL=FALSE)\n",
    "if (length(gset) > 1) idx <- grep(\"GPL6246\", attr(gset, \"names\")) else idx <- 1\n",
    "gset <- gset[[idx]]\n",
    "\n",
    "# group names for all samples in a series\n",
    "gsms <- \"10101010\"\n",
    "sml <- c()\n",
    "for (i in 1:nchar(gsms)) { sml[i] <- substr(gsms,i,i) }\n",
    "sml <- paste(\"G\", sml, sep=\"\")  set group names\n",
    "\n",
    "# order samples by group\n",
    "ex <- exprs(gset)[ , order(sml)]\n",
    "sml <- sml[order(sml)]\n",
    "fl <- as.factor(sml)\n",
    "labels <- c(\"prueba\",\"control\")\n",
    "\n",
    "# set parameters and draw the plot\n",
    "palette(c(\"#dfeaf4\",\"#f4dfdf\", \"#AABBCC\"))\n",
    "dev.new(width=4+dim(gset)[[2]]/5, height=6)\n",
    "par(mar=c(2+round(max(nchar(sampleNames(gset)))/2),4,2,1))\n",
    "title <- paste (\"GSE18388\", '/', annotation(gset), \" selected samples\", sep ='')\n",
    "boxplot(ex, boxwex=0.6, notch=T, main=title, outline=FALSE, las=2, col=fl)\n",
    "legend(\"topleft\", labels, fill=palette(), bty=\"n\")\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Pregunta 2: Non-Alcoholic Fatty Liver Disease\n",
    "\n",
    "Elegimos el número de acceso GSE72756 debido a que plantea el estudio de las *Long non-codig RNA*, cuya influencia sobre la enfermedad analizada aún no ha sido determinada, recayendo en un estudio interesante de esta. El organismo claramente es el *Homo Sapiens*\n",
    "\n",
    "![](NAFLDGEOSeries.png)\n",
    "\n",
    "Como ya realizamos los pasos anteriormente determinamos los grupos `prueba` y `control` una vez más para experimentar:\n",
    "\n",
    "![](GSE72756.png)\n",
    "\n",
    "![](GruposNAFLD.png)\n",
    "\n",
    "Obtuvimos lo siguiente:\n",
    "\n",
    "![](ResultadosNAFLDTabla.png)\n",
    "\n",
    "El rango de los valores-P sin procesar es $[0.0000033,0.005243]$.\n",
    "\n",
    "Acá están los resultados para los 4 primeros genes.\n",
    "\n",
    "![](ResultadosNAFLDG1.png)\n",
    "\n",
    "![](ResultadosNAFLDG2.png)\n",
    "\n",
    "![](ResultadosNAFLDG3.png)\n",
    "\n",
    "![](ResultadosNAFLDG4.png)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.3.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
